Update Kubernetes dependency versions to 1.20.6

AxiomSamarth opened the pull request:
**What this PR does / why we need it**:
This PR vendors K8s 1.20.5

**Which issue(s) this PR fixes**:
Fixes #402 

**Special notes for your reviewer**:
The PR is divided into 3 commits
- [Commit 1](https://github.com/gardener/machine-controller-manager/pull/601/commits/7b448428c53bb7e8b6bc0525a89f62b5a972d047) - Update go.mod to use k8s v1.20.5
- [Commit 2](https://github.com/gardener/machine-controller-manager/pull/601/commits/db79470a9b97ec859f27a2c6bf60378994588571) - Vendored files for k8s v1.20.5
- [Commit 3](https://github.com/gardener/machine-controller-manager/pull/601/commits/a9964f35de323909ac7eaef86be1499ce7e363b5) - Introducing contexts for every controller and cascading them across

Right now, all the contexts are `context.Background()`. We need to understand the contract between the provider APIs and contexts before introducing the context based timeouts in creation/deletion timeouts.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```other user
Update Kubernetes dependency versions to v1.20.6
```

prashanth26 commented on pkg/controller/alicloudmachineclass.go:
```diff
@@ -193,7 +194,7 @@ func (c *controller) reconcileClusterAlicloudMachineClass(class *v1alpha1.Aliclo
 			// If controller has deleteMigratedMachineClass flag set
 			// and the migratedMachineClass annotation is set
-			err = c.controlMachineClient.AlicloudMachineClasses(class.Namespace).Delete(class.Name, &metav1.DeleteOptions{})
+			err = c.controlMachineClient.AlicloudMachineClasses(class.Namespace).Delete(context.TODO(), class.Name, metav1.DeleteOptions{})
```
You will need to determine the right way to pass the context here. I think `context.TODO()` is a hack. Ideally we will need some sort of timeout. However, probably need a global context as well from which we derive contexts. 

You will need to research more about it. Something [like this](https://github.com/gardener/gardener/blob/77e09dd522d7bfb682cc35d0be437201b606061b/pkg/utils/kubernetes/shoot.go#L47-L50)?

prashanth26 reviewed (changes requested):
Also the commits look messed up. Can we fix that?

AxiomSamarth commented:
Important change that will be shipped along with this vendoring - 

The [`cache.waitForCacheSync()`](https://github.com/kubernetes/client-go/blob/05b7f154ae6b5138e1ca4dfbbbdce7a01d8992b9/tools/cache/shared_informer.go#L254) in the new Kubernetes v1.20.5 has the method of [`wait.PollImmediateUntil()`](https://github.com/kubernetes/apimachinery/blob/dd12c7a65e7f49c4e00b5bd5eb93de2fcbf7831d/pkg/util/wait/wait.go#L499) which runs the 'condition' before waiting for the interval and it somehow fails to update the caches properly (at least in our codebase). 

So, a new package has been introduced as `/pkg/utils/provider/cache` which copies the older implementation of `waitForCacheSync()` with [`wait.PollUntil()`](https://github.com/kubernetes/apimachinery/blob/dd12c7a65e7f49c4e00b5bd5eb93de2fcbf7831d/pkg/util/wait/wait.go#L489) which always waits interval before the first run of 'condition'.

AxiomSamarth commented:
Here is a note regarding a very typical issue commonly seen with MacOS which then leads to build failure at `check` point with the error 

```
pull-request-gardener.machine-controller-manager-pr.master/.ci/check
build github.com/googleapis/gnostic/openapiv2: cannot load github.com/googleapis/gnostic/openapiv2: open /tmp/build/d56d9853/pull-request-gardener.machine-controller-manager-pr.master/vendor/github.com/googleapis/gnostic/openapiv2: no such file or directory
```

`vendor/github.com/googleapis/gnostic/OpenAPIv2` couldn't be located/identified despite being vendored because the name of the package is of mixed case and MacOS is by default case insensitive. What helped? Below -

- delete the vendored file with mixedcase using `git rm` cmd
- run go mod vendor again and then push

```
git rm -rf vendor/github.com/googleapis/gnostic/OpenAPIv2
rm 'vendor/github.com/googleapis/gnostic/OpenAPIv2/OpenAPIv2.go'
rm 'vendor/github.com/googleapis/gnostic/OpenAPIv2/OpenAPIv2.pb.go'
rm 'vendor/github.com/googleapis/gnostic/OpenAPIv2/OpenAPIv2.proto'
rm 'vendor/github.com/googleapis/gnostic/OpenAPIv2/README.md'
rm 'vendor/github.com/googleapis/gnostic/OpenAPIv2/openapi-2.0.json'
```

```
@C02XR0WNJGH5 machine-controller-manager % go mod vendor
```

```
@C02XR0WNJGH5 machine-controller-manager % git status
On branch k8s-io-dep-1.20.5
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	deleted:    vendor/github.com/googleapis/gnostic/OpenAPIv2/OpenAPIv2.go
	deleted:    vendor/github.com/googleapis/gnostic/OpenAPIv2/OpenAPIv2.pb.go
	deleted:    vendor/github.com/googleapis/gnostic/OpenAPIv2/OpenAPIv2.proto
	deleted:    vendor/github.com/googleapis/gnostic/OpenAPIv2/README.md
	deleted:    vendor/github.com/googleapis/gnostic/OpenAPIv2/openapi-2.0.json
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   go.sum
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	vendor/github.com/googleapis/gnostic/openapiv2/
```

AxiomSamarth commented:
The PR is divided into five commits.

[Commit 1](https://github.com/AxiomSamarth/machine-controller-manager/tree/6bd96cef04ec7b122e54e029de3d1b5c4e25ad51) - Update the go.mod file with the versions relevant for k8s v1.20.5

[Commit 2](https://github.com/AxiomSamarth/machine-controller-manager/tree/f72fc2c5a44429c06058128c48fd5da1e84cc941) - This commit only has the vendors files

[Commit 3](https://github.com/AxiomSamarth/machine-controller-manager/tree/a50638e6d1de96e7a4eead7dafc8f4c91d933b52) - This commit has all the changes in the core MCM packages aligned with the K8s v1.20.5

[Commit 4](https://github.com/AxiomSamarth/machine-controller-manager/tree/c87f515d29f50224007026c7d35f98736a744e0f) - This is the vendoring of openapiv2 due to the issue mentioned in the [comment](https://github.com/gardener/machine-controller-manager/pull/601#issuecomment-866570065) 

Commit 5 - to be pushed soon and it shall introduce timeout based contexts on kubernetes rest clients.

prashanth26 commented on pkg/controller/alicloudmachineclass.go:
```diff
@@ -215,44 +217,44 @@ func (c *controller) reconcileClusterAlicloudMachineClass(class *v1alpha1.Aliclo
 	clone := class.DeepCopy()
 	clone.Finalizers = finalizers
-	_, err = c.controlMachineClient.AlicloudMachineClasses(class.Namespace).Update(clone)
+	_, err = c.controlMachineClient.AlicloudMachineClasses(class.Namespace).Update(context.TODO(), clone, metav1.UpdateOptions{})
```
```suggestion
	_, err = c.controlMachineClient.AlicloudMachineClasses(class.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
```

prashanth26 commented on pkg/controller/awsmachineclass.go:
```diff
@@ -215,44 +217,44 @@ func (c *controller) reconcileClusterAWSMachineClass(class *v1alpha1.AWSMachineC
 	clone := class.DeepCopy()
 	clone.Finalizers = finalizers
-	_, err = c.controlMachineClient.AWSMachineClasses(class.Namespace).Update(clone)
+	_, err = c.controlMachineClient.AWSMachineClasses(class.Namespace).Update(context.TODO(), clone, metav1.UpdateOptions{})
```
```suggestion
	_, err = c.controlMachineClient.AWSMachineClasses(class.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
```
Check other such files

prashanth26 commented on pkg/util/provider/drain/drain.go:
```diff
@@ -635,6 +635,7 @@ func (o *Options) checkAndDeleteWorker(volumeAttachmentEventCh chan *storagev1.V
 }
 
 func (o *Options) evictPodsWithPVInternal(
```
I suspect this function or it's wrapping function has a mainContext created and passed around. With this change now, we can rather use this background context to derive the context?

prashanth26 reviewed (commented):
I skimmed through (as there are too many files with changes) the changes. Looks fine to me. 

- Just search through all non `_test.go` files to see if you have mised anyother context changes.
- In the drain handling we could now use the passed context instead of creating a new maint context. 
- Finally, I think it's okay to not introduce the timeouts in this PR. However we will need to introduce this change to address [such stories](https://github.com/gardener/machine-controller-manager/issues/494#issuecomment-844923912)

prashanth26 commented:
/hold until MCM v.40.0 is released.

AxiomSamarth commented:
> I skimmed through (as there are too many files with changes and even github is getting stuck with the number of files 🗡️ ) the changes. Looks fine to me.
> 
> * In the drain handling we could now use the passed context instead of creating a new main context.

I guess, you mean we could use the passed context `ctx` to create the `mainContext` with deadline. That is, to use below code

```
mainContext, cancelMainContext = context.WithDeadline(ctx, o.drainStartedOn.Add(o.Timeout))
defer cancelMainContext()
```

instead of

```
mainContext, cancelMainContext = context.WithDeadline(context.Background(), o.drainStartedOn.Add(o.Timeout))
defer cancelMainContext()
```

Please tell me if I understood it right? Because, if we replace `mainContext `with `ctx` then that would wipe out the essence of `mainContext` for the reason it is present [here](https://github.com/AxiomSamarth/machine-controller-manager/blob/a9964f35de323909ac7eaef86be1499ce7e363b5/pkg/util/provider/drain/drain.go#L651) in `drain.go`

prashanth26 commented:
> Please tell me if I understood it right? Because, if we replace mainContext with ctx then that would wipe out the essence of mainContext for the reason it is present here in drain.go

Yes, this is what I meant.

prashanth26 commented on pkg/controller/awsmachineclass_test.go:
```diff
@@ -121,21 +122,21 @@ var _ = Describe("machineclass", func() {
 
 				action := data.action
-				machineClass, err := controller.controlMachineClient.AWSMachineClasses(TestNamespace).Get(action.machineClassName, metav1.GetOptions{})
+				machineClass, err := controller.controlMachineClient.AWSMachineClasses(TestNamespace).Get(context.TODO(), action.machineClassName, metav1.GetOptions{})
```
I would think we need to pass a bg.context here in this file as well.

AxiomSamarth replied:
Most of the tests have `context.TODO(`) set. So, should be changed across all tests to `context.Background()`?

prashanth26 replied:
Sorry, my bad. I mistook it to be for `pkg/controller/awsmachineclass`. The test files can continue to have context.TODO() I think. 

Ignore my previous comment. The page loading was too slow, hence got confused I think.

prashanth26 reviewed (approved):
I just skimmed throught the files as there are too many changes. As long as there is no logic change and works well with local testing. 

/lgtm just one minor comment

amshuman-kr commented on pkg/controller/machine_safety_test.go:
```diff
@@ -290,18 +291,48 @@ var _ = Describe("#machine_safety", func() {
+			// Validate MachineSets with Setup Data
 			Expect(len(machineSets)).To(Equal(1))
 
+			for annotation := range data.setup.machineSetAnnotations {
+				Expect(data.setup.machineSetAnnotations[annotation]).To(Equal(machineSets[0].Annotations[annotation]))
+			}
+
+			for label := range data.setup.machineSetLabels {
+				Expect(data.setup.machineSetLabels[label]).To(Equal(machineSets[0].Labels[label]))
+			}
+
+			if len(machineSets[0].Status.Conditions) > 0 {
+				Expect(data.setup.machineSetConditions[0]).To(Equal(machineSets[0].Status.Conditions[0]))
+			}
+
+			// Validate MachineDeployment with the Setup Data
+			machineDeployments, err := c.machineDeploymentLister.List(labels.Everything())
+			Expect(err).To(BeNil())
+
+			for annotation := range data.setup.machineDeploymentAnnotations {
+				Expect(data.setup.machineDeploymentAnnotations[annotation]).To(Equal(machineDeployments[0].Annotations[annotation]))
+			}
+
+			for label := range data.setup.machineDeploymentLabels {
+				Expect(data.setup.machineDeploymentLabels[label]).To(Equal(machineDeployments[0].Labels[label]))
+			}
+
+			if len(machineDeployments[0].Status.Conditions) > 0 {
+				Expect(data.setup.machineDeploymentConditions[0]).To(Equal(machineDeployments[0].Status.Conditions[0]))
+			}
+
```
```suggestion
			for k, v := range data.setup.machineSetAnnotations {
				Expect(machineSets[0].Annotations).To(HaveKeyWithValue(k, v))
			}

			for k, v := range data.setup.machineSetLabels {
				Expect(machineSets[0].Labels).To(HaveKeyWithValue(k, v))
			}

			if len(machineSets[0].Status.Conditions) > 0 {
				Expect(machineSets[0].Status.Conditions).To(ContainElements(data.setup.machineSetConditions...))
			}

			// Validate MachineDeployment with the Setup Data
			machineDeployments, err := c.machineDeploymentLister.List(labels.Everything())
			Expect(err).To(BeNil())

			for k, v := range data.setup.machineDeploymentAnnotations {
				Expect(machineDeployments[0].Annotations).To(HaveKeyWithValue(k, v))
			}

			for k, v := range data.setup.machineDeploymentLabels {
				Expect(machineDeployments[0].Labels).To(HaveKeyWithValue(k, v))
			}

			if len(machineDeployments[0].Status.Conditions) > 0 {
				Expect(machineDeployments[0].Status.Conditions).To(ContainElements(data.setup.machineDeploymentConditions...))
			}

```

AxiomSamarth replied:
Adopted this suggestion with a minor change of typecasting the matcher to type `[]v1.MachineDeploymentConditions` instead of variable length interface.

amshuman-kr commented on pkg/util/provider/app/app.go:
```diff
@@ -306,7 +306,7 @@ func getAvailableResources(clientBuilder coreclientbuilder.ClientBuilder) (map[s
 
 		healthStatus := 0
-		resp := client.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
+		resp := client.Discovery().RESTClient().Get().AbsPath("/healthz").Do(context.TODO()).StatusCode(&healthStatus)
```
Shouldn't this be `context.Background()` or context passed as an argument to the function?

AxiomSamarth replied:
You're right, Amshu. This function is referred within `StartControllers`. So, the context need not be passed as argument as there is no propagated usage. 

This should have been `context.Background()`. I will make the change.

amshuman-kr reviewed (commented):
Sorry for the delay in reviewing. There were far too many files in the PR :-)

The changes look good. Just a few minor comments below.

AxiomSamarth commented on go.mod:
```diff
@@ -1,55 +1,51 @@
-	k8s.io/code-generator => k8s.io/code-generator v0.0.0-20190912054826-cd179ad6a269 // kubernetes-1.16.0
-	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20190816220812-743ec37842bf
+	k8s.io/api => k8s.io/api v0.20.5
+	k8s.io/apimachinery => k8s.io/apimachinery v0.20.6-rc.0
```
This version was fetched & set automatically when I ran `go mod vendor` as corresponding release for `k8s.io/api v0.20.5`. I would like to understand if there is a way to validate the same.

AxiomSamarth replied:
I have set the version to v0.20.6 across all packages as we discussed.

AxiomSamarth commented on go.mod:
```diff
@@ -1,55 +1,51 @@
+	k8s.io/apiserver => k8s.io/apiserver v0.20.5
+	k8s.io/client-go => k8s.io/client-go v0.20.5
+	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.20.5
+	k8s.io/code-generator => k8s.io/code-generator v0.20.6-rc.0
```
This version was fetched & set automatically when I ran `go mod vendor` as corresponding release for `k8s.io/api v0.20.5`. I would like to understand if there is a way to validate the same.

prashanth26 commented:
PR is failing due to some linting issues. You can check the pipeline or run `make check/verify` to list them locally.

amshuman-kr reviewed (approved):
Thanks for the changes @AxiomSamarth! LGTM

prashanth26 commented:
/title Update Kubernetes dependency versions to 1.20.6
